synapse/federation/sender/__init__.py

```diff
@@ -560,3 +577,35 @@ async def get_replication_rows(
         # Dummy implementation for case where federation sender isn't offloaded
         # to a worker.
         return [], 0, False
+
+    async def _wake_destinations_needing_catchup(self):
+        """
+        Wakes up destinations that need catch-up and are not currently being
+        backed off from.
+        Does so in a slow way (one every 5 seconds) to reduce load spikes.
+        """
+
+        last_processed = None  # type: Optional[str]
+
+        while True:
+            destinations_to_wake = await self.store.get_catch_up_outstanding_destinations(
+                last_processed
+            )
+
+            if not destinations_to_wake:
+                # finished waking all destinations!
+                break
+
+            for destination in destinations_to_wake:
+                if self._federation_shard_config.should_handle(
+                    self._instance_name, destination
+                ):
+                    last_processed = destination
+                    logger.info(
+                        "Destination %s has outstanding catch-up, waking up.",
+                        destination,
+                    )
+                    self.wake_destination(destination)
```

Comment:
Is it possible that `destination` no longer needs to be woken up at this point or will `wake_destination` just bail since the loop is running already?

Reply:
If it already started transmitting and is doing that, it will see the flag and no do anything more.

If it already caught up quickly, then it will still wake it up but the logic there will notice that there is nothing to send so it won't do anything.